feat: Set values from local files

mumoshu opened the pull request:
This adds the `--set-file key=filepath` flag to `install`, `upgrade`, `template` and `lint` sub-commands so that the content of the file at the `filepath` is set to the value for the `key`.

Resolves #1754

cwedgwood reviewed (commented):
lgtm, i will try this tomorrow

mumoshu commented:
Hi @technosophos @bacongobbler @adamreese 

Would you mind reviewing this? This is an implementation for the feature requested since one and a half years 😉

mumoshu commented:
@cwedgwood Thanks for taking your time!

mumoshu commented:
Fixed bugs and tests after a self-review.

mumoshu commented:
Just realized that I have to also cover `helm lint`. Just a moment...

mumoshu commented:
Done.

mumoshu commented:
Missed doc regen 😉Done.

davidkarlsen commented:
@cwedgwood Any progress on this?

davidkarlsen commented:
@cwedgwood @technosophos can you please have a look at the PR?

cwedgwood commented:
@davidkarlsen it's not my call ... given there are nearly 100 merge requests pending, I imagine this might take some time.

davidkarlsen commented:
@cwedgwood OK - sorry for the nagging - who would be the most appropriate reviewer/merger?

cwedgwood commented:
@davidkarlsen i think what you've done is fine/appropriate; given the backlog of issues and pull requests you might want to consider forking for now (or merging into another fork that suits you better)

davidkarlsen commented:
I'm just hunting for the feature and there seems to be a fair share of interest for it, it is actually @mumoshu doing the hard work :)

OK, we'll sit in line and watch the queue :)

bacongobbler reviewed (changes requested):
Hey @mumoshu! I had a look at this and the code looks great. Would you mind adding a note about this new feature flag under the following sections?

https://github.com/kubernetes/helm/blob/48e703997016f3edeb4f0b90e6cfdb3456ce3db0/docs/chart_best_practices/values.md#consider-how-users-will-use-your-values

https://github.com/kubernetes/helm/blob/bae4f4bb248a114a0d6cd4fec62d98f5f5453ded/docs/using_helm.md#customizing-the-chart-before-installing

If you find any other sections where --set-file might be useful, please also feel free to add it into the docs. We just need to document new features; we've heard from users that they want to make sure all new features coming into Helm are documented with clear usage examples.

Thanks so much!

bacongobbler commented:
@davidkarlsen you rang? ;)

Since this PR is a `size/L` we'll also need the sign-off from another maintainer.

mumoshu commented:
@bacongobbler Thanks for the review. Good point! I completed adding docs to everywhere I think good.

Would you mind taking a look once again?

mumoshu commented:
Just to make it sure - please feel free to squash everything into a single commit on github when you merge. Of course in any case you want me to do so, just tell me so! Thanks

bacongobbler reviewed (approved):
All very clear examples. Nice touch on adding it to `helm help install`. Thanks!

davidkarlsen commented:
We need another reviewer, maybe @technosophos ?

mumoshu commented:
@bacongobbler Would you mind poking me when you're about to merge this?
I'll try to rebase this asap then :)

davidkarlsen commented:
/assign @technosophos

dheerajjoshim commented:
This isn't available in 2.10.0-rc.1 ?
https://github.com/kubernetes/helm/releases/tag/v2.10.0-rc.1

mumoshu commented:
@adamreese Hi! Would it be possible to ask you to merge this?

mumoshu committed:
feat: Set values from local files

Adds the `--set-file key=filepath` flag to `install`, `upgrade`, `template` and `lint` sub-commands so that the content of the file at the `filepath` is set to the value for the `key`.

Resolves #1754

mumoshu commented:
Rebased and squashed commits.

technosophos commented on cmd/helm/lint.go:
```diff
@@ -47,10 +45,15 @@ type lintCmd struct {
+
+	certFile string
+	keyFile  string
+	caFile   string
```
Why were TLS options added to Lint, which is explicitly forbidden from making server connections?

mumoshu replied:
True! I don't remember why I did it but I'll remove it asap. Thx.

mumoshu replied:
Turned out it was due to my misunderstanding. Reverted the tls options and the relevant refactoring altogether - It should be good to go now!

The longer explanation follows.

Firstly, I removed the `vals` func dedicated to the `lint` command and migrated it to reuse another `vals` func used across other commands like `upgrade`.

While migrating,  I believe that I thought it would be better to add support for tls opts for the `vals` func, for use when downloading remote value files. And it resulted in adding those flags. But as you pointed out, it conflicts with the design of the `lint` command, that is forbidden from making server connections. I reverted those changes and it should be good now :)

mumoshu commented:
@technosophos Thanks for the review! I reverted the changes that turned out wrong. It should be good to go now.

mumoshu commented:
For your reference, [here's the commits I made](https://github.com/mumoshu/helm/compare/daa0d07...set-value-from-local-file-before-squash) in order to revert things. They're already incorporated and squashed into this PR, but I thought it would help your review.

technosophos reviewed (approved):
Thank you for the quick turn-around on the TLS options. This is looking great.

wpc009 commented:
The [document](https://docs.helm.sh/chart_best_practices/values/#consider-how-users-will-use-your-values) is not very clear about how this `--set-file` option works. 
Finally I figured it out that `--set-file key=path_to_file` will  treat the content of `path_to_file` as multilines string, and set it to `key`. This description should be added in the docs.
The first time I see the `--set-file` option, I expect that helm will import values from the file provided.

mumoshu commented:
@wpc009 Hey! Thx for the feedback.
Honestly I did my best to make it as clear as possible.
So would you mind proposing alternative sentence(s)?
